### PR TITLE
Utility revision

### PR DESCRIPTION
**Pull request by unpluggedk**
Update base64 encoding/decoding
Refactor hashing in NSData and NSString
Refactor to have a correct initializers in right classes and convenience methods

Add missing unit tests

Includes work from closed PRs #149 and #144

### CODE REVIEW

File: IdentityCore/src/util/NSString+MSIDExtensions.h [hunk 1]
```diff
@@ -42,26 +46,48 @@
...
++ (NSString *)msidRandomUrlSafeStringOfByteSize:(NSUInteger)size;
 
-- (BOOL)msidIsEquivalentWithAnyAlias:(NSArray<NSString *> *)aliases;
+/*! Generate a URL-safe string of random data */
++ (NSString *)msidHexStringFromData:(NSData *)data;
+
+/*! Generate a base64 url-encoded string of random data */
++ (NSString *)msidBase64UrlEncodedStringFromData:(NSData *)data;
+
+/*! Generate a www-form-urlencoded string of random data */
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.h]: the comment is incorrect here

File: IdentityCore/src/util/NSString+MSIDExtensions.h [hunk 2]
```diff
@@ -42,26 +46,48 @@
...
+/*!
+ =============================================================================
+ Convenience methods
+ =============================================================================
+ */
+/*! Convenience method to convert string to url */
++ (NSString *)msidStringFromOrderedSet:(NSOrderedSet *)set;
+
+/*! Convenience method to convert string to url */
+- (NSURL *)msidUrl;
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.h]: do we really need this?

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 3]
```diff
@@ -71,208 +38,45 @@ @implementation NSString (MSIDExtensions)
...
-    
-    if(!pbEncoded){
-        return nil;
+    NSData *data = [self.class msidBase64UrlDecodeData:self];
+    if (!data) return nil;
+    
+    char lastByte;
+    [data getBytes:&lastByte range:NSMakeRange([data length] - 1, 1)];
+    
+    // We need to check for null terminated string data by looking at the last bit.
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: if there's any Apple documentation referencing this behavior, would be great to add a link

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 4]
```diff
@@ -403,10 +160,80 @@ + (NSString *)randomUrlSafeStringOfSize:(NSUInteger)size
...
+/// </summary>
+/// <remarks>
+/// See RFC 4648, Section 5 plus switch characters 62 and 63 and no padding.
+/// For a good overview of Base64 encoding, see http://en.wikipedia.org/wiki/Base64
+/// This SDK will use rfc7515 and encode without using padding.
+/// See https://tools.ietf.org/html/rfc7515#appendix-C
+/// </remarks>
++ (NSString *)msidBase64UrlEncodedStringFromData:(NSData *)data
+{
+    return [[data base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithLineFeed] msidStringByRemovingPadding];
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: Just curious, why NSDataBase64EncodingEndLineWithLineFeed?

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 5]
```diff
@@ -403,10 +160,80 @@ + (NSString *)randomUrlSafeStringOfSize:(NSUInteger)size
...
 
 
+/*! Generate a www-form-urlencoded string of random data */
++ (NSString *)msidUrlFormEncodedStringFromDictionary:(NSDictionary *)dict
+{
+    __block NSMutableString *encodedString = nil;
+    
+    [dict enumerateKeysAndObjectsUsingBlock: ^(id key, id value, BOOL __unused *stop)
+     {
+         NSString *encodedKey = [[[key description] msidTrimmedString] msidUrlFormEncode];
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: Why do we use "[key description]" and not key?

Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: Should we skip appending key, value if encodedKey is nil?

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 6]
```diff
@@ -403,10 +160,80 @@ + (NSString *)randomUrlSafeStringOfSize:(NSUInteger)size
...
+         if (!encodedString)
+         {
+             encodedString = [NSMutableString new];
+         }
+         else
+         {
+             [encodedString appendString:@"&"];
+         }
+         
+         [encodedString appendFormat:@"%@", encodedKey];
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: Why not use appendString?

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 7]
```diff
@@ -424,4 +251,48 @@ - (BOOL)msidIsEquivalentWithAnyAlias:(NSArray<NSString *> *)aliases
...
+
++ (NSString *)msidStringFromOrderedSet:(NSOrderedSet *)set
+{
+    NSInteger cSet = set.count;
+    if (cSet == 0)
+    {
+        return @"";
+    }
+    
+    NSMutableString *queryString = [[set objectAtIndex:0] mutableCopy];
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: nit: might be easier to use componentsJoinedByString here with a tradeoff of using nsarray

File: IdentityCore/src/util/NSString+MSIDExtensions.m [hunk 8]
```diff
@@ -424,4 +251,48 @@ - (BOOL)msidIsEquivalentWithAnyAlias:(NSArray<NSString *> *)aliases
...
+}
+
+
+- (NSOrderedSet<NSString *> *)msidScopeSet
+{
+    return [NSOrderedSet msidOrderedSetFromString:self];
+}
+
+
+- (NSString *)msidStringByRemovingPadding
```
Comment (oldalton) [inline IdentityCore/src/util/NSString+MSIDExtensions.m]: nit: I'd prefer this method to specify that it removes base64 padding, not any padding

File: IdentityCore/tests/MSIDOrderedSetExtensionsTests.m [hunk 9]
```diff
@@ -0,0 +1,48 @@
+//
+//  MSIDOrderedSetExtensionsTests.m
+//  IdentityCore
+//
+//  Created by Jason Kim on 9/4/18.
```
Comment (oldalton) [inline IdentityCore/tests/MSIDOrderedSetExtensionsTests.m]: change header

File: IdentityCore/src/util/NSDictionary+MSIDExtensions.m [hunk 10]
```diff
@@ -78,72 +78,35 @@ + (NSDictionary *)msidDictionaryFromString:(NSString *)string
     return queryDict;
 }
 
-// Encodes a dictionary consisting of a set of name/values pairs that are strings to www-form-urlencoded
-// Returns nil if the dictionary is empty, otherwise the encoded value
-- (NSString *)msidURLFormEncode
++ (NSDictionary *)msidDictionaryFromData:(NSData *)data error:(NSError **)error
```
[the remark this answers is not in the document]
Comment (unpluggedk) [reply IdentityCore/src/util/NSDictionary+MSIDExtensions.m]: Done

File: IdentityCore/src/util/NSString+MSIDExtensions.m (hunk 6 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: I agree! Brought this over and didn't change this.

File: IdentityCore/src/util/NSString+MSIDExtensions.m (hunk 4 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: I don't see a strong reason to have any mask here actually. 
Thus, putting 0 .

File: IdentityCore/src/util/NSString+MSIDExtensions.h (hunk 2 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.h]: Moved it to test util, as we have quite a few cases which uses this.

File: IdentityCore/src/util/NSString+MSIDExtensions.m (hunk 5 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: Sure, key is always string :)

Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: Sure

File: IdentityCore/src/util/NSString+MSIDExtensions.m (hunk 7 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: good point!

File: IdentityCore/src/util/NSString+MSIDExtensions.m (hunk 8 above)
Comment (unpluggedk) [reply IdentityCore/src/util/NSString+MSIDExtensions.m]: I actually got rid of this. it was only being used at one place and it is a one liner.

File: IdentityCore/tests/MSIDOrderedSetExtensionsTests.m (hunk 9 above)
Comment (unpluggedk) [reply IdentityCore/tests/MSIDOrderedSetExtensionsTests.m]: done :)

**Pull request merged**